### ISSUE TITLE
Document different types of estimation

### ISSUE THREAD

**Issue by johnrichardrinehart** [infracost/infracost#1663]
I ran `infracost` on a directory of `terraform` files and the below was printed near the end of the output
```bash
* 48 were estimated, 33 of which include usage-based costs, see https://infracost.io/usage-file
```
How were the costs estimated for the 48-33=15 resources which did _not_ include usage-based costs (i.e. what else, besides usage can be used to estimate the cost of cloud resources which don't have a fixed rate cost based on time)? Also, how can I determine _which_ cloud resources correspond to these 15?

This behavior should be documented - it might be, already. If it's not documented, already, then if someone answers the above questions then I'm happy to PR the documentation if I'm also told which file(s) should be modified to include this information. :pray:

**Comment by alikhajeh1** [comment]
Hi @johnrichardrinehart - thanks for creating this. I can see where the confusion is coming from... The remaining 15 resources only had fixed rate cost components. They're essentially the ones that have **no** "cost depends on usage" line items, so for example the following would be a non-usage-based-resource:
```
Project: myproject

 Name                                                                           Monthly Qty  Unit                        Monthly Cost
 module.metric_alarm_target_4xx.aws_cloudwatch_metric_alarm.this[0]
 └─ Standard resolution                                                                   1  alarm metrics                      $0.10
```

It would be helpful if you could describe what CLI output you expected to see, and why you want that? The context helps a lot :) 

#### Infracost JSON output
As an aside, the [Infracost JSON output](https://www.infracost.io/docs/features/cli_commands/#examples) has a `summary` field that has the total resource counts but it only has the resource type/count list for unsupportedResourceCounts and noPriceResourceCounts (free resources). It doesn't include the resource types for the totalSupportedResources as the CLI text output shows the list, but maybe you're looking for a summary?

```
$ infracost breakdown --path . --terraform-parse-hcl --show-skipped --format json | jq ".summary"

{
  "totalDetectedResources": 79,
  "totalSupportedResources": 21,
  "totalUnsupportedResources": 0,
  "totalUsageBasedResources": 12,
  "totalNoPriceResources": 58,
  "unsupportedResourceCounts": {},
  "noPriceResourceCounts": {
    "aws_db_parameter_group": 1,
    "aws_db_subnet_group": 1,
    "aws_eip": 1,
    "aws_iam_openid_connect_provider": 1,
    "aws_iam_policy": 3,
    "aws_iam_role": 7,
    "aws_iam_role_policy_attachment": 9,
    "aws_internet_gateway": 1,
    "aws_lambda_permission": 2,
    "aws_route": 2,
    "aws_route53_record": 1,
    "aws_route_table": 2,
    "aws_route_table_association": 9,
    "aws_security_group": 3,
    "aws_security_group_rule": 6,
    "aws_sns_topic_subscription": 1,
    "aws_subnet": 7,
    "aws_vpc": 1
  }
}
```

**Comment by johnrichardrinehart** [comment]
@alikhajeh1 Thanks for your response! I'm sure you answered my question but I'm still not fully understanding :disappointed: .

> It would be helpful if you could describe what CLI output you expected to see, and why you want that? The context helps a lot :)

I expect to see:
1) an hourly/monthly cost for those resources whose cost can be determined without requiring an amount of usage (GB of egress, number of queries executed, number of Cloud Function triggers, etc.)
and
2) an estimated cost for those resources whose cost can only be estimated knowing both a) the price and b) an amount tof usage (the usage being provided in `usage-file.yaml`, say).

And, I have those two things (great job `infracost` developers :smile: ) . But, then I have this _extra_ thing which is the "estimated cost for those things whose cost does _not_ depend on an amount of usage". It's beyond what I expected to see and I'm not exactly sure what it is. So, when you say

> The remaining 15 resources only had fixed rate cost components. They're essentially the ones that have no "cost depends on usage" line items

Are you saying that the cost is _not_ an estimate (emphasis for readability, with no tone implied)? The way I see it (and this could be my misunderstanding), either

1) The cost needs to be estimated because the cost depends both on a) the price and b) some unknown usage.
2) The cost is known because the cost is fixed and does not depend on an amount of usage (i.e. these resources have fixed costs).

So, a component with a fixed rate cost does not seem to require a cost estimate. So, what am I missing?

The only resolution that I can devise is that the cost is an estimate (and not a known quantity) because
a) "fixed rate cost" means that the resource's cost is based on time that the resource is "alive" (provisioned)
and
b) the resource is assumed to have been "alive" (provisioned) for the whole month (or hour, depending on if it's an hourly cost estimate or monthly cost estimate) when, in reality, the resource may only have been provisioned for a few days within the month or a few minutes within an hour.

Is my above guess (at the resolution) correct?

**Comment by alikhajeh1** [comment]
> Is my above guess (at the resolution) correct?

Yep, that's a detailed and correct description of what Infracost does :)

>  I'm only looking for an understanding of what it means to estimate cost for resources whose cost does not depend on usage.

Sorry my bad, I should have explained it in a clearer way. Infracost multiplies the usage by the price to calculate a cost estimate - there is no other magic. Thus the CLI:
1. parses the usage values out of the Terraform code where possible (e.g. disk size)
2. uses the usage-file values (e.g. for lambda requests)
3. uses 730 hours per month (e.g. for instance usage)

The order and exact logic of this depends on the resource that is being estimated, e.g. [here](https://github.com/infracost/infracost/blob/master/internal/resources/aws/lambda_function.go#L33-L108) is the code for Lambda. But regardless of that, the Monthly Qty/Unit and Price columns show the values being used in the calculation. Running `infracost breakdown --fields all` shows the Price column too.

We always use the word "estimate" as things like reserved instances and price discounts usually apply to the bill (which shows the actual cost).

There's an [open issue](https://github.com/infracost/infracost/issues/1556) to make the 730 hours customizable for resources such as instances.

Does that help?

**Comment by johnrichardrinehart** [comment]
Oh, I think I understand! Sorry for the long back-and-forth.

So, basically "estimate" outside of "usage-based costs" means "assume a certain amount of usage time". So, when I see something like

```
338 cloud resources were detected:
∙ 47 were estimated, 32 of which include usage-based costs, see https://infracost.io/usage-file
∙ 281 were free, rerun with --show-skipped to see details
∙ 10 are not supported yet, rerun with --show-skipped to see details
```

Then that means that the `OVERALL TOTAL` is calculated by adding up the costs of
a) 32 resources whose cost may be incorrect because the inferred usage (from `usage-file.yml`, say)  is incorrect/unavailable
b) 15 resources whose (monthly) cost does _not_ depend on 3rd-party usage but _does_ depend on the total amount of time that the resource has been alive/provisioned during the past 730 hours.

Okay, so now a follow-up question and a follow-up statement and a follow-up suggestion that I hope may ultimately result in an improvement in the reporting interface/details.

## The Question

Are there resources that do _not_ depend on total time provisioned/active? Think "you pay a fixed/atomic cost at provision-time. If so, then the distinction between usage-based and time-based is insufficient and, probably, another pricing model should be spat out in the summary.

## The Statement

**Comment by alikhajeh1** [comment]
@johnrichardrinehart thanks - that distinction is helpful. The usage-file currently contains some time-usage (e.g. [this](https://github.com/infracost/infracost/blob/master/infracost-usage-example.yml#L451-L454)) and we have future plans to move [all params](https://github.com/infracost/infracost/issues/1525) into the usage-file (meaning that anyone could generate a cost estimate, even without having terraform). That might render the distinction less useful as the usage-file could be edited to customize the entire estimate, or do what-if analysis. Does that make sense?

But even then, I might be useful to see the distinction you highlight as that's a question users have early on, before they know what a usage-file is (or that they can also [fetch usage from CloudWatch](https://www.infracost.io/docs/features/usage_based_resources/#fetch-from-cloudwatch)).

**Comment by johnrichardrinehart** [comment]
> Does that make sense?

100%.

> But even then, I might be useful to see the distinction you highlight as that's a question users have early on

Yeah, I think it might be best to call them "cost calculations" instead of "cost estimates". The calculations may be based on values like "alive time" and "peripheral usage" and "resource configuration values". Some of these values may need to be estimated if they aren't known/provided.

Also, would you mind answering the question from my [earlier message](https://github.com/infracost/infracost/issues/1663#issuecomment-1128874230)?

**Comment by alikhajeh1** [comment]
Re your question, I can't think of any.
What it comes down to is that when we say "usage-based costs" what we really mean is "the costs that are effected by things that can't be determined from the IaC alone" (for example, the size of an S3 bucket or the number of Lambda requests). You make a good point that we don't state that directly, so it may be good to define that in [this page](https://www.infracost.io/docs/features/usage_based_resources/).

**Comment by johnrichardrinehart** [comment]
> what we really mean is "the costs that are effected by things that can't be determined from the IaC alone"

I actually think that's not what you mean. Because, time is a resource that's not captured by the IaC. That's my point. "usage" can mean "the amount of time for which the resource is used". This is why I think it's a good idea for `infracost` to disambiguate between "time-based cost", "peripheral-based cost", and "configuration-based cost" components.

**Comment by glenngillen** [comment]
We're moving to a contribution model where Issues will be exclusively for work that is ready to be worked on. Discussions and feature requests about potential future capabilities will be over in [Discussions](https://github.com/infracost/infracost/discussions). There has been very little community or customer interest in this change for an extended period, but if it still interests you (or you're now reading this issue because you've been looking for a solution) please [create a new thread in Discussions](https://github.com/infracost/infracost/discussions).
